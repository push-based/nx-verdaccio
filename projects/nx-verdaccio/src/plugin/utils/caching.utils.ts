import { hashObject } from 'nx/src/hasher/file-hasher';

export function cacheKey(prefix: string, hashData: Record<string, unknown>) {
  return `${prefix}-${hashObject(hashData)}`;
}
