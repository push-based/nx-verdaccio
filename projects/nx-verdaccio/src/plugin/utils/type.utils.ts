export type Prettify<T> = {
  [K in keyof T]: T[K];
} & {};

export type WithRequired<T, K extends keyof T> = Prettify<
  Required<Pick<T, K>> & Omit<T, K>
>;
