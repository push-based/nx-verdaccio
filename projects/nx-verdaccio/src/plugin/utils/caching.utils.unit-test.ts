import {
  afterEach,
  beforeEach,
  describe,
  expect,
  type MockInstance,
} from 'vitest';
import { cacheKey } from './caching.utils';

import * as fileHasher from 'nx/src/hasher/file-hasher';

describe('cacheKey', (): void => {
  const prefix = 'warcraft';
  const hashData = { race: 'orc' };
  const hashObjectReturnValue = '123456789';
  const regex = /^[a-zA-Z]+-\d+$/;

  let hashObjectSpy: MockInstance<[obj: object], string>;

  beforeEach((): void => {
    hashObjectSpy = vi
      .spyOn(fileHasher, 'hashObject')
      .mockReturnValue(hashObjectReturnValue);
  });
  afterEach((): void => {
    hashObjectSpy.mockRestore();
  });

  it('should return cache key with unmodified prefix', (): void => {
    expect(cacheKey(prefix, {})).toBe(`${prefix}-${hashObjectReturnValue}`);
  });

  it('should return a value in the format "string-numbers"', (): void => {
    expect(cacheKey(prefix, hashData)).toMatch(regex);
  });

  it('should call hashObject once, and with correct argument', (): void => {
    cacheKey(prefix, hashData);
    expect(hashObjectSpy).toHaveBeenCalledTimes(1);
    expect(hashObjectSpy).toHaveBeenCalledWith(hashData);
  });

  it('should return cache key, when hashData is empty', (): void => {
    expect(cacheKey(prefix, {})).toBe(`${prefix}-${hashObjectReturnValue}`);
  });

  it('should return cache key, when hashData is NOT empty', (): void => {
    expect(cacheKey(prefix, hashData)).toBe(
      `${prefix}-${hashObjectReturnValue}`
    );
  });
});
