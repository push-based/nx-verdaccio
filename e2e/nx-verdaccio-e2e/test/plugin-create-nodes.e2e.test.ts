import type { Tree } from '@nx/devkit';
import { join } from 'node:path';
import { afterEach, expect } from 'vitest';
import {
  addJsLibToWorkspace,
  materializeTree,
  nxShowProjectJson,
  registerPluginInWorkspace,
} from '@push-based/test-nx-utils';
import { updateProjectConfiguration } from 'nx/src/generators/utils/project-configuration';
import {
  TARGET_ENVIRONMENT_BOOTSTRAP,
  TARGET_ENVIRONMENT_SETUP,
  TARGET_PACKAGE_INSTALL,
  TARGET_PACKAGE_PUBLISH,
  TARGET_ENVIRONMENT_VERDACCIO_START,
  TARGET_ENVIRONMENT_INSTALL,
  TARGET_ENVIRONMENT_VERDACCIO_STOP,
} from '@push-based/nx-verdaccio';
import { teardownTestFolder } from '@push-based/test-utils';
// eslint-disable-next-line @nx/enforce-module-boundaries
import {
  TARGET_ENVIRONMENT_E2E,
  TARGET_ENVIRONMENT_TEARDOWN,
} from '../../../projects/nx-verdaccio/src/plugin/targets/environment.targets';

describe('nx-verdaccio plugin create-nodes-v2', () => {
  let tree: Tree;
  const projectA = 'lib-a';
  const projectB = 'lib-b';
  const projectAE2e = `${projectA}-e2e`;
  const e2eProjectARoot = join('projects', projectAE2e);
  const baseDir = `tmp/environments/${process.env['NX_TASK_TARGET_PROJECT']}/__test__/create-nodes-v2`;

  beforeEach(async () => {
    tree = await addJsLibToWorkspace(projectA);
    await addJsLibToWorkspace(projectB, tree);
    await addJsLibToWorkspace(projectAE2e, tree);
    updateProjectConfiguration(tree, projectAE2e, {
      root: e2eProjectARoot,
      projectType: 'application',
    });
  });

  afterEach(async () => {
    // await teardownTestFolder(baseDir);
  });

  it('should add package targets to library project', async () => {
    const cwd = join(baseDir, 'add-pkg-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, projectA);
    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({
      [TARGET_PACKAGE_INSTALL]: expect.objectContaining({
        dependsOn: [
          {
            target: TARGET_PACKAGE_PUBLISH,
            params: 'forward',
          },
          {
            target: TARGET_PACKAGE_INSTALL,
            projects: 'dependencies',
            params: 'forward',
          },
        ],
        executor: '@push-based/nx-verdaccio:pkg-install',
      }),
      [TARGET_PACKAGE_PUBLISH]: expect.objectContaining({
        dependsOn: [
          {
            params: 'forward',
            target: 'build',
          },
          {
            params: 'forward',
            projects: 'dependencies',
            target: 'nxv-pkg-publish',
          },
        ],
        executor: '@push-based/nx-verdaccio:pkg-publish',
      }),
    });

    expect(projectJson.targets).toMatchSnapshot();
  });

  it('should NOT add package targets to application project', async () => {
    const cwd = join(baseDir, 'no-pkg-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    await materializeTree(tree, cwd);

    const { projectJson } = await nxShowProjectJson(cwd, projectAE2e);

    expect(projectJson.targets).toStrictEqual(
      expect.not.objectContaining({
        [TARGET_PACKAGE_INSTALL]: expect.any(Object),
        [TARGET_PACKAGE_PUBLISH]: expect.any(Object),
      })
    );
  });

  it('should add package targets to library project if some tag of options.packages.filterByTag match', async () => {
    const cwd = join(baseDir, 'add-pkg-targets-filterByTag');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
        packages: {
          filterByTags: ['publish'],
        },
      },
    });
    updateProjectConfiguration(tree, projectB, {
      root: `projects/${projectB}`,
      sourceRoot: 'projects/lib-b/src',
      projectType: 'library',
      tags: ['publish'],
    });
    await materializeTree(tree, cwd);

    const { projectJson: projectJsonB } = await nxShowProjectJson(
      cwd,
      projectB
    );
    expect(projectJsonB.name).toBe(projectB);
    expect(projectJsonB.tags).toStrictEqual(['publish']);
    expect(projectJsonB.targets).toStrictEqual(
      expect.objectContaining({
        [TARGET_PACKAGE_INSTALL]: expect.any(Object),
        [TARGET_PACKAGE_PUBLISH]: expect.any(Object),
      })
    );

    const { projectJson: projectJsonA } = await nxShowProjectJson(
      cwd,
      projectA
    );
    expect(projectJsonA.name).toBe(projectA);
    expect(projectJsonA.tags).toStrictEqual([]);
    expect(projectJsonA.targets).toStrictEqual(
      expect.not.objectContaining({
        [TARGET_PACKAGE_INSTALL]: expect.any(Object),
        [TARGET_PACKAGE_PUBLISH]: expect.any(Object),
      })
    );
  });

  it('should add environment targets to project with targetName e2e dynamically', async () => {
    const cwd = join(baseDir, 'add-env-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    updateProjectConfiguration(tree, projectAE2e, {
      root: e2eProjectARoot,
      projectType: 'application',
      targets: {
        e2e: {},
      },
    });
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, projectAE2e);
    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual(
      expect.objectContaining({
        e2e: expect.objectContaining({
          configurations: {},
          dependsOn: [
            {
              params: 'forward',
              target: TARGET_ENVIRONMENT_SETUP,
            },
          ],
        }),
        [TARGET_ENVIRONMENT_BOOTSTRAP]: expect.objectContaining({
          executor: '@push-based/nx-verdaccio:env-bootstrap',
        }),
        [TARGET_ENVIRONMENT_INSTALL]: expect.objectContaining({
          dependsOn: [
            {
              params: 'forward',
              projects: 'dependencies',
              target: TARGET_PACKAGE_INSTALL,
            },
          ],
          executor: 'nx:run-commands',
          options: {
            environmentRoot: expect.toMatchPath('tmp/environments/lib-a-e2e'),
            command: expect.stringContaining(
              'echo "dependencies installed for'
            ),
          },
        }),
        [TARGET_ENVIRONMENT_SETUP]: expect.objectContaining({
          cache: true,
          executor: '@push-based/nx-verdaccio:env-setup',
          options: {},
          inputs: [
            '{projectRoot}/project.json',
            {
              runtime: 'node --version',
            },
            {
              runtime: 'npm --version',
            },
            {
              externalDependencies: ['verdaccio'],
            },
            '^production',
          ],
          outputs: [
            '{options.environmentRoot}/.npmrc',
            '{options.environmentRoot}/package.json',
            '{options.environmentRoot}/package-lock.json',
            '{options.environmentRoot}/node_modules',
          ],
        }),
        [TARGET_ENVIRONMENT_VERDACCIO_START]: expect.objectContaining({
          executor: '@nx/js:verdaccio',
          options: expect.objectContaining({
            clear: true,
            config: '.verdaccio/config.yml',
            environmentDir: expect.toMatchPath('tmp/environments/lib-a-e2e'),
            port: expect.any(Number), // random port number
            projectName: 'lib-a-e2e',
            storage: expect.toMatchPath('tmp/environments/lib-a-e2e/storage'),
          }),
        }),
        [TARGET_ENVIRONMENT_VERDACCIO_STOP]: expect.objectContaining({
          executor: '@push-based/nx-verdaccio:kill-process',
          options: {
            filePath: expect.toMatchPath(
              'tmp/environments/verdaccio-registry.json'
            ),
          },
        }),
        [TARGET_ENVIRONMENT_E2E]: expect.objectContaining({
          executor: '@push-based/nx-verdaccio:env-teardown',
          dependsOn: [
            {
              params: 'forward',
              target: 'e2e',
            },
          ],
        }),
        [TARGET_ENVIRONMENT_TEARDOWN]: expect.objectContaining({
          executor: '@push-based/nx-verdaccio:env-teardown',
        }),
      })
    );
  });

  it('should NOT add environment targets to project without targetName e2e', async () => {
    const cwd = join(baseDir, 'no-env-targets');
    registerPluginInWorkspace(tree, {
      plugin: '@push-based/nx-verdaccio',
      options: {
        environments: {
          targetNames: ['e2e'],
        },
      },
    });
    await materializeTree(tree, cwd);

    const { projectJson } = await nxShowProjectJson(cwd, projectAE2e);

    expect(projectJson.targets).toStrictEqual(
      expect.not.objectContaining({
        [TARGET_ENVIRONMENT_BOOTSTRAP]: expect.any(Object),
        [TARGET_ENVIRONMENT_INSTALL]: expect.any(Object),
        [TARGET_ENVIRONMENT_SETUP]: expect.any(Object),
        [TARGET_ENVIRONMENT_VERDACCIO_START]: expect.any(Object),
        [TARGET_ENVIRONMENT_VERDACCIO_STOP]: expect.any(Object),
      })
    );
  });
});
