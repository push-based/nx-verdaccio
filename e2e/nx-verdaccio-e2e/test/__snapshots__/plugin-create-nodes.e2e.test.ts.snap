// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nx-verdaccio plugin create-nodes-v2 > should add package targets to library project 1`] = `
{
  "nxv-pkg-install": {
    "configurations": {},
    "dependsOn": [
      {
        "params": "forward",
        "target": "nxv-pkg-publish",
      },
      {
        "params": "forward",
        "projects": "dependencies",
        "target": "nxv-pkg-install",
      },
    ],
    "executor": "@push-based/nx-verdaccio:pkg-install",
    "options": {},
    "parallelism": true,
  },
  "nxv-pkg-publish": {
    "configurations": {},
    "dependsOn": [
      {
        "params": "forward",
        "target": "build",
      },
      {
        "params": "forward",
        "projects": "dependencies",
        "target": "nxv-pkg-publish",
      },
    ],
    "executor": "@push-based/nx-verdaccio:pkg-publish",
    "options": {},
    "parallelism": true,
  },
}
`;
